Ecosystem: Refactor AppCard for different tag colours

javabster opened the pull request:
## Changes

This PR refactors the AppCard component to accept objects as well as arrays for the tags. This then allows us to edit the colours of the tags as needed for the ecosystem page

based on discussions from #2638 

UPDATE: closes #2681 

cc @1ucian0 @IceKhan13

## Screenshots
![screencapture-localhost-3000-ecosystem-2022-06-10-16_29_50](https://user-images.githubusercontent.com/23662430/173146572-f31c85c7-ff05-4bb3-ba85-7365fb4f34e7.png)

javabster commented:
@JRussellHuffman I used the default 'purple' option that comes with the carbon tag (which i believe is accessible by default). Happy to change the styles if you have a preferred colour for branding/accessibility purposes :)

techtolentino commented:
@JRussellHuffman 

The colors for the text + label, that @javabster introduced pass the color contrast tests, using the [WCAG Extension [1]](https://chrome.google.com/webstore/detail/wcag-color-contrast-check/plnahcmalebffmaghcpcmpaciebdhgdf?hl=en) in Chrome, as well as the [Color Contrast Checker tool [2]](https://marijohannessen.github.io/color-contrast-checker/), recommended by Carbon.

1) ![Screen Shot 2022-06-14 at 10 36 58 AM (2)](https://user-images.githubusercontent.com/6276074/173646259-aa2c46c6-ff3b-4510-bec3-140377f236be.png)

2) ![Screen Shot 2022-06-14 at 10 40 06 AM (2)](https://user-images.githubusercontent.com/6276074/173647081-f88f1d67-f33d-4ff5-9878-aa89d9d7b1ca.png)

techtolentino commented on components/ui/AppCard.vue:
```diff
@@ -17,12 +17,20 @@
+            v-for="tag in Object.values(tags).flat()"
+            :key="tag"
+            :label="tag"
+            :kind="tag === tags.tier ? 'purple' : 'dark-purple'"
```
@javabster - please check the [Carbon Vue tag component](https://vue.carbondesignsystem.com/?path=/info/components-cvtag--standard), to see if this is allowed... I think `purple` is a valid value, however, I suspect that `dark-purple` is causing issues in the console:

![Screen Shot 2022-06-14 at 11 09 37 AM (2)](https://user-images.githubusercontent.com/6276074/173659543-87347cdc-8342-4919-a026-bd2bb8f751bc.png)


we might have to leverage one of the pre-existing colors?

javabster commented:
thanks for the comments @techtolentino! I've addressed your comments and now theres no console warnings when I run locally 😄

techtolentino commented:
Hey @JRussellHuffman 

Earlier in our discussion, we talked about leaving the app card tags as all dark purple. However, after speaking with @javabster, it turns out that there _is_ a need to distinguish between different tag tiers... and that users need to be able to differentiate.

I'm just commenting here to start the discussion, but @javabster will have more info.

thanks!

javabster commented:
Just to add a bit more clarity on the motivations for this PR: 

After discussion with @1ucian0 and @IceKhan13 we thought it was confusing to have the 'tier' and 'repo label' tags on the AppCard to be in the same dark purple colour (similar to what was discussed in #2638). We thought we should either a) make the 'tier' tags a different colour or b) figure out a different way to display the 'tier' on the AppCard other than as a tag. I opened this PR using approach (a) as it was technically simpler to implement, not realising it had design implications due to the colour naming restrictions from the carbon component.

JRussellHuffman commented:
There are a few problems with just changing the colors.
1. accessibility. Rule of thumb, color cannot be the primary carrier of information because colorblind users wouldn't be able to tell the difference between the tags. In other words, color can't be the only difference between the cards if they mean different things, but adding an icon (for example) would solve for this requirement
2. UX. This goes hand in hand with the accessibility requirement, but if I'm a user on the page and the only difference between the tags is color, it's unclear to me how they are different.

It sounds like the "b" option of figuring out a different way to display the 'tier' might be better here. But I can help more if you can share some details about what the tiers are.

eddybrando commented:
**Note:** We're currently waiting for the resolution of an offline discussion to get a final UI suggestion. Until then, we shouldn't merge this PR since we haven't decided _how_ we want to handle these cases.

FYI: @JRussellHuffman @javabster @techtolentino

techtolentino commented:
I'm ok w/ closing this PR @eddybrando

eddybrando commented:
Hi @javabster, just to confirm: We've decided to implement the following UI, right?

![Screenshot 2022-07-01 at 13 58 27](https://user-images.githubusercontent.com/22047320/176890189-d6037c53-e74e-4691-ba39-cdcd9b90c249.png)

Source: https://www.figma.com/file/N34caXM86bY1778Whk1c31/2503-ecosystem-page?node-id=307%3A8021

javabster commented:
@eddybrando yes that's the plan currently!

But now that #2681 has come up, i'm actually wondering if it would be better to go with this option instead and kill 2 birds with one stone? 

![image](https://user-images.githubusercontent.com/23662430/176945087-7527eb07-9026-4f9a-bf1f-24e8ac13763a.png)

@1ucian0 do you have a preference?

1ucian0 commented:
Yeap. Looks good to me😊 thank you!

On Fri, Jul 1, 2022, 7:46 PM Abby Mitchell ***@***.***> wrote:

> @eddybrando <https://github.com/eddybrando> yes that's the plan currently!
>
> But now that #2681 <https://github.com/Qiskit/qiskit.org/issues/2681> has
> come up, i'm actually wondering if it would be better to go with this
> option instead and kill 2 birds with one stone?
>
> [image: image]
> <https://user-images.githubusercontent.com/23662430/176945087-7527eb07-9026-4f9a-bf1f-24e8ac13763a.png>
>
> @1ucian0 <https://github.com/1ucian0> do you have a preference?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Qiskit/qiskit.org/pull/2660#issuecomment-1172578212>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAF3FZIRJEDXUISJNT7XJD3VR4VHTANCNFSM5YO2CJZA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

javabster commented:
After some internal discussion we have decided to implement the following design, this will also close #2681 

![image](https://user-images.githubusercontent.com/23662430/177820837-205998ba-1752-4368-8f5b-47edd6890241.png)

y4izus reviewed (changes requested):
Hey Abby!! Thank for working on this!! I have some comments, but let's start first with the visual and behavioral part. After fixing that, I would like to do another review with more focus on the code itself :)

- On medium and small screens, there is a misalignment between the title of the card and the tags in the advocates page (where we use the same component)

Now:
![Captura de Pantalla 2022-07-11 a las 13 33 52](https://user-images.githubusercontent.com/17231966/178255339-3a382867-3ae0-443c-8a00-8033a47dc38e.png)

With your change:
![Captura de Pantalla 2022-07-11 a las 13 33 32](https://user-images.githubusercontent.com/17231966/178255285-8cdbd7c0-7216-4763-a054-d7ef8cadbc7a.png)

- On the filters section, I only see "Main" and "MAIN". I thought we should have all our tiers
![Captura de Pantalla 2022-07-11 a las 13 37 37](https://user-images.githubusercontent.com/17231966/178255944-d49bdfac-49db-4ee2-9229-6c948d5ed1ab.png)

- I think the text on the tooltip is missing
![Captura de Pantalla 2022-07-11 a las 13 38 30](https://user-images.githubusercontent.com/17231966/178256118-a638d66e-469c-4571-9158-a3e145809c05.png)

- There are two different tags for "community", should we have just one?
![Captura de Pantalla 2022-07-11 a las 13 40 30](https://user-images.githubusercontent.com/17231966/178256390-1038ca05-41ab-4403-8106-ed35dcc77c38.png)
![Captura de Pantalla 2022-07-11 a las 13 40 40](https://user-images.githubusercontent.com/17231966/178256419-1232fc99-8e31-46c9-ad30-aeaf7c378b61.png)

- On the [figma file](https://www.figma.com/file/N34caXM86bY1778Whk1c31/2503-ecosystem-page?node-id=307%3A8021) content on the tags start with uppercase and the rest is lower case. on this PR we have both everything on lower case or everything on uppercase. Example:
![Captura de Pantalla 2022-07-11 a las 13 42 35](https://user-images.githubusercontent.com/17231966/178256731-9af721c3-0740-410b-be92-58d0dbc88c5d.png)

- Filter is case sensitive. When I select "Main" I don't get any card
![Captura de Pantalla 2022-07-11 a las 13 46 21](https://user-images.githubusercontent.com/17231966/178257348-45216e96-349e-4c60-bce9-355aa5afdffd.png)
I don't get any message neither... @JRussellHuffman , @techtolentino , if it's not done, could you define what happened when the filter doesn't return any result? Thank you!!

javabster commented:
Hey Yaiza thanks for your comments! I'll get working on the changes, just a few responses:

- Yep the filler text is still there bc I'm waiting on some changes in the backend before I can update, I probably should have changed this to a draft PR as it's not technically finished yet sorry 😅 
- Can you rerun `npm run generate` and see if you still see the capitalisation issue? I think you may have an outdated version of the `memebrs.json` as I don't see this issue when I run locally
- I think we need to keep the 2 different 'community' tags as one refers to a Tier (which has specific criteria attached to it) and the other is just a more general tag that the project owners have the option to set for their repo. When i get the tooltip text hopefully this distinction will be clearer 😄

y4izus commented:
You're right Abby, I forgot to run `npm run generate`, sorry!!

javabster commented:
Hey @y4izus I made a few changes based on your feedback:

- removed semicolons ✅ 

- fixed margin on purple tags:
<img width="246" alt="Screenshot 2022-07-15 at 12 37 09 PM" src="https://user-images.githubusercontent.com/23662430/179295636-9e28cfe4-40d6-4ff5-ab36-7a1140b5533a.png">

- I'm just waiting on [this](https://github.com/qiskit-community/ecosystem/pull/178) ecosystem PR to get merged then i can update the tooltip filler text

lmk if there's anything else 😄

javabster commented:
<img width="732" alt="Screenshot 2022-07-21 at 3 38 53 PM" src="https://user-images.githubusercontent.com/23662430/180301270-1c76bc2a-034d-4854-bfe0-30b51dd59f21.png">

Tooltips now work! This is ready for a final review I think @y4izus

eddybrando commented:
The one UX issue that I have with this implementation is that I believe that the tiers' descriptions would be more useful to the user before starting to filter, so that they could know what they are filtering for.

With this approach, the description is only seen in the results but not in the filter control elements. To know what the filters refer to in detail, one would need to first find every different tier tag in the cards and hover over the tooltips.

<img width="1590" alt="Screenshot 2022-08-01 at 14 23 37" src="https://user-images.githubusercontent.com/22047320/182146970-bf95c062-48ee-4e6d-a878-88b20ef38527.png">

eddybrando reviewed (changes requested):
Thank you very much @javabster 🙂 

I have two minor UI-related comments:

---

1. The tags should be aligned to the right and not have this space:

<img width="636" alt="image" src="https://user-images.githubusercontent.com/22047320/182148213-a2b1c331-a925-482e-9bbf-fd6a11b89814.png">

---

2. These spacing seems off to me. It feels like the tags need some bottom margin, but I couldn't find a Figma design with this layout, where the tags are aligned to the left on certain breakpoints:

<img width="408" alt="image" src="https://user-images.githubusercontent.com/22047320/182148592-7b77f424-8d08-4b25-9dfe-76bda5b42617.png">

The designs that I found [in Figma](https://www.figma.com/file/N34caXM86bY1778Whk1c31/2503-ecosystem-page?node-id=64%3A3953) only show a version with inline tags aligned to the right:

<img width="139" alt="Screenshot 2022-08-01 at 14 36 34" src="https://user-images.githubusercontent.com/22047320/182149128-02b6f66b-13e1-4719-ad9c-a71789bd67e5.png">

@JRussellHuffman could you please help us clarify how it should look like?

eddybrando commented:
> @eddybrando @JRussellHuffman
> 
> 1. Sure I'll adjust the spacing 👍
> 2. For the tag spacing on the line below this is how it's always been due to the way the AppCard component was built. you can see the same effect for the Advocates page currently in prod. I can change it if you're happy with changing the underlying behaviour of the component, or if it requires more thought as it will affect multiple pages perhaps it could be put into a followup issue?
> 
> As for the tooltip I was following the design from Russell that we agreed upon [a number of weeks ago](https://github.com/Qiskit/qiskit.org/pull/2660#issuecomment-1177869093). We need some way to indicate on the app card itself the difference between the tier tags and the general tags. Perhaps we could add the tooltip to the filters section in addition to keeping them here?

Thank you! 🙂

Thanks for pointing out that the current design of the AppCard already has the spacing "issue" that I mentioned. There is no need to make any further change to that in this PR as it is out of scope.

And yes, I'm also happy with moving forward with the implementation of what was agreed upon. It's definitely a big improvement and valuable to the user. I just wanted to share some UX concerns I have, but we could implement them later if we think they bring any additional value.

javabster commented:
@eddybrando I adjusted the spacing! lmk if theres anything else needed on this 😄 


<img width="610" alt="Screenshot 2022-08-04 at 10 50 52 AM" src="https://user-images.githubusercontent.com/23662430/182883472-50d03a33-8190-4893-ae71-bf618c7119b4.png">

y4izus reviewed (changes requested):
Thank you Abby!! I have some comments about the code. Also, could you double check the design of the card? I see the license not aligned
<img width="251" alt="Captura de Pantalla 2022-08-08 a las 13 06 44" src="https://user-images.githubusercontent.com/17231966/183404404-95ed4fc4-d092-4eab-8c3c-d7bf4ec19e6f.png">

Lastly, could you check the console too?? I'm getting a lot of warnings
<img width="793" alt="Captura de Pantalla 2022-08-08 a las 13 07 55" src="https://user-images.githubusercontent.com/17231966/183404599-74d0273f-ab54-4cf0-aeaf-a9d7c769864e.png">

javabster commented:
Thanks @y4izus I've addressed your comments 😄 I also added a bit more margin under the tags as i was editing the app card component anyway, so this also addresses @eddybrando comment from earlier

Latest version looks like the following when i run locally, I don't see the spacing issue or any errors in the console:

<img width="1350" alt="Screenshot 2022-08-08 at 2 56 23 PM" src="https://user-images.githubusercontent.com/23662430/183493963-3dbfd3b7-850d-4dee-ba57-92358618c197.png">
<img width="488" alt="Screenshot 2022-08-08 at 2 56 39 PM" src="https://user-images.githubusercontent.com/23662430/183493967-25e5b8e7-8647-47ed-b6be-29ec137d9e98.png">

y4izus commented on components/ui/AppCard.vue:
```diff
@@ -201,3 +207,56 @@ export default class AppCard extends Vue {
+    display: inline-flex;
+    min-width: 0;
+    max-width: 100%;
+    min-height: 1.5rem;
```
Please, use for margin and padding the carbon spacing variables like you did [here](https://github.com/Qiskit/qiskit.org/pull/2660/files#diff-39a1c5dd7d7f4b0c976425ad09d8aae1069e5758cbe7a1f616550c1dc1e4ef1eR242)

y4izus commented on components/ui/AppCard.vue:
```diff
@@ -201,3 +207,56 @@ export default class AppCard extends Vue {
+      fill: white;
+      margin-left: $spacing-02;
+
+    :hover svg {
```
Could you add the spaces/tabs needed to this block, is it inside `.bx--tooltip__trigger svg`, right?

y4izus commented on components/ui/AppCard.vue:
```diff
@@ -70,6 +86,10 @@ export default class AppCard extends Vue {
     return Array.isArray(this.tags) && this.tags.length > 0
   }
+
+  get hasTooltipTags () {
```
`hasTags` and `hasTooltipTags` do the same with different variables, you can create an unique method and reuse it passing it the variable

y4izus reviewed (changes requested):
Thank you very much Abby!! Just a few more small modifications and we are done!! 👏

y4izus commented on pages/ecosystem/index.vue:
```diff
@@ -168,6 +170,16 @@ import { GeneralLink } from '~/constants/appLinks'
+
+    getTierDescription (tier: string) {
+      const tip = (this as any).tiers.find((tip: any) => tip.name === tier)
+      return tip.description
```
Sorry, I've forgotten this one. Please, add an option just in case there are no description: `return tip.description || ""` , for example

y4izus reviewed (approved):
LGTM!!!! Thank you Abby!!! 👏
